example-raft-kv/src/rpc_handlers.rs

```diff
@@ -0,0 +1,114 @@
+use std::collections::BTreeSet;
+
+use actix_web::get;
+use actix_web::post;
+use actix_web::web;
+use actix_web::web::Data;
+use actix_web::Responder;
+use openraft::raft::AppendEntriesRequest;
+use openraft::raft::ClientWriteRequest;
+use openraft::raft::EntryPayload;
+use openraft::raft::InstallSnapshotRequest;
+use openraft::raft::VoteRequest;
+use openraft::NodeId;
+use serde::Deserialize;
+use serde::Serialize;
+use web::Json;
+
+use crate::app::ExampleApp;
+use crate::store::ExampleRequest;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Empty {}
+
+// --- Raft communication
+
+#[post("/raft-vote")]
+pub async fn vote(app: Data<ExampleApp>, req: Json<VoteRequest>) -> actix_web::Result<impl Responder> {
+    let res = app.raft.vote(req.0).await;
+    Ok(Json(res))
+}
+
+#[post("/raft-append")]
+pub async fn append(
+    app: Data<ExampleApp>,
+    req: Json<AppendEntriesRequest<ExampleRequest>>,
+) -> actix_web::Result<impl Responder> {
+    let res = app.raft.append_entries(req.0).await;
+    Ok(Json(res))
+}
+
+#[post("/raft-snapshot")]
+pub async fn snapshot(app: Data<ExampleApp>, req: Json<InstallSnapshotRequest>) -> actix_web::Result<impl Responder> {
+    let res = app.raft.install_snapshot(req.0).await;
+    Ok(Json(res))
+}
+
+// --- Application API
+
+#[post("/write")]
+pub async fn write(app: Data<ExampleApp>, req: Json<ExampleRequest>) -> actix_web::Result<impl Responder> {
+    let res = app.raft.client_write(ClientWriteRequest::new(EntryPayload::Normal(req.0))).await;
+    Ok(Json(res))
+}
+
+#[post("/read")]
+pub async fn read(app: Data<ExampleApp>, req: Json<String>) -> actix_web::Result<impl Responder> {
+    let res = {
+        let sm = app.store.sm.read().await;
+        let key = req.0;
+        let value = sm.kvs.get(&key).cloned();
+        value.unwrap_or_default()
+    };
+    Ok(Json(res))
+}
```

Comment:
I think this could result in stale reads if the request reaches a node that is not the leader, right?

IMHO an endpoint with potential stale reads could be very useful in many applications, but it should be explicit that it could be a stale read. Adding another endpoint based on `client_read` would be good.

Reply:
Good point!
This is a draft application.
I'm gonna add a read endpoint in another PR:DDD

Reply:
@MikaelCall @drmingdrmer Is this outdated? By placing the code below before all state machine calls will sort this issue?

```rust
let _ = app.raft.client_read().await;
```

I suspect it will wait until the node gets confirmed to be in sync with all nodes, no?

Reply:
@ppamorim 
You have to handle the error. If the node that receives the request is not the leader it will give a forward to leader error. Not handling the error (i.e. returning the error using ?) may still result in stale reads.

Search for client_read in this repository. The fn docs are good and there is a great test explaining the difference calling `client_read` on leader vs follower nodes.

Reply:
@MikaelCall Nice! I was confused and thought it was part of another code, in my project I removed this completely and it's redirecting the request to the leader, as expected. I will read more about it too.